File: c++/2096.cpp
#include <algorithm>
#include <iostream>

using namespace std;

int arr[100001][3];
int pre[2][3];
int curr[2][3];

int main() {
  freopen("example.txt", "r", stdin);

  int n;
  cin >> n;

  for (int i = 1; i <= n; i++) {
    cin >> arr[i][0] >> arr[i][1] >> arr[i][2];
  }

  for (int i = 1; i <= n; i++) {
    curr[0][0] = arr[i][0] + max(pre[0][0], pre[0][1]);
    curr[0][1] = arr[i][1] + max({pre[0][0], pre[0][1], pre[0][2]});
    curr[0][2] = arr[i][2] + max(pre[0][1], pre[0][2]);

    curr[1][0] = arr[i][0] + min(pre[1][0], pre[1][1]);
    curr[1][1] = arr[i][1] + min({pre[1][0], pre[1][1], pre[1][2]});
    curr[1][2] = arr[i][2] + min(pre[1][1], pre[1][2]);

    for (int j = 0; j < 2; j++) {
      for (int k = 0; k < 3; k++) {
        pre[j][k] = curr[j][k];
      }
    }
  }

  int max = curr[0][0];
  int min = curr[1][0];

  for (int i = 1; i < 3; i++) {
    if (max < curr[0][i]) max = curr[0][i];
    if (min > curr[1][i]) min = curr[1][i];
  }

  cout << max << ' ' << min;

  return 0;
}

File: js/2096.js
// js는 메모리 초과로 통과가 안 됨.
const fs = require('fs');

const inputs = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const n = +inputs[0];
const arr = [[0, 0, 0]];

for (let i = 1; i <= n; i++) {
  const subArr = inputs[i].split(' ').map(Number);
  arr.push(subArr);
}

let pre = Array(2)
  .fill()
  .map(() => Array(3).fill(0));

const curr = Array(2)
  .fill()
  .map(() => Array(3).fill(0));

for (let i = 1; i <= n; i++) {
  curr[0][0] = arr[i][0] + Math.max(pre[0][0], pre[0][1]);
  curr[0][1] = arr[i][1] + Math.max(pre[0][0], pre[0][1], pre[0][2]);
  curr[0][2] = arr[i][2] + Math.max(pre[0][1], pre[0][2]);

  curr[1][0] = arr[i][0] + Math.min(pre[1][0], pre[1][1]);
  curr[1][1] = arr[i][1] + Math.min(pre[1][0], pre[1][1], pre[1][2]);
  curr[1][2] = arr[i][2] + Math.min(pre[1][1], pre[1][2]);

  pre = [...curr].map((arr) => [...arr]);
}

console.log(Math.max(...curr[0]), Math.min(...curr[1]));
